exchange/exchange.go

```diff
@@ -936,6 +936,88 @@ func (e *exchange) getBidCacheInfo(bid *pbsOrtbBid, auction *auction) (cacheInfo
 	return
 }
 
+// getAuctionCurrencyRates returns the constant currency conversion rates that will be used to process this
+// auction. Parameter customRates contains, if any, the bidRequest custom currency conversion rates
+// defined in bidRequest.Ext.  If usePbsRates flag is set to false, all previous Conversions values are discarded
+// and this function merely substitutes Conversions with customRates. But if usePbsRates was ommitted or set to
+// true, customRates' values update or add to the Conversions map
+func (e *exchange) getAuctionCurrencyRates(customRates *openrtb_ext.ExtRequestCurrency) currency.Conversions {
+	// No bidRequest.ext custom rates defined, use PBS rates as usual
+	if customRates == nil || len(customRates.ConversionRates) == 0 {
+		return e.currencyConverter.Rates()
+	}
+
+	// bidRequest.ext defines custom rates of its own, figure out whether to use a mix of
+	// Pbs rates and custom rates or custom rates only
+	if customRates.UsePBSRates != nil && !(*customRates.UsePBSRates) {
+		// Use custom rates only
+		includeInverseRates(customRates.ConversionRates)
+
+		return currency.NewRates(time.Time{}, customRates.ConversionRates)
+	} else {
+		// Use both custom and PBS rates
+		pbsConversions := e.currencyConverter.Rates()
+		serverRates := pbsConversions.GetRates()
+		auctionRates := make(map[string]map[string]float64, len(*serverRates))
+
+		copyCustomRates(auctionRates, customRates.ConversionRates)
```

Comment:
Wouldn't it be easier here if we just looped through all the custom rates and override the values in PBS rates (or a copy of PBS rates) that need to be overridden? Something like:
```
// Go through all the custom rates
for fromCur, rates := range customRates {
	for toCur, val := range rates {
		// If the rate conversion exists in PBS rates then override the value
		if _, ok := pbsRates[fromCur]; ok {
			pbsRates[fromCur][toCur] = val
		} else if _, ok := pbsRates[toCur]; ok {
			pbsRates[toCur][fromCur] = val
		} else {
			// If it doesn't exist then, add the conversion
			pbsRates[fromCur][toCur] = val
		}
	}
}
```

(This code snippet is just to give you an idea of what I was thinking. May or may not compile as is)

Reply:
I considered that option at first but discarded in favor of the new code found inside the newly created `currency/rate_engines.go`. Let me know your take on this.